src/Npgsql/Replication/ReplicationConnection.cs

```diff
@@ -816,7 +820,9 @@ void SetTimeouts(TimeSpan readTimeout, TimeSpan writeTimeout)
         {
             var connector = Connector;
             connector.UserTimeout = readTimeout > TimeSpan.Zero ? (int)readTimeout.TotalMilliseconds : 0;
-            connector.WriteBuffer.Timeout = writeTimeout;
+
+            if (connector.WriteBuffer != null)
+                connector.WriteBuffer.Timeout = writeTimeout;
```

Comment:
```suggestion
            connector.WriteBuffer?.Timeout = writeTimeout;
```

Reply:
Does that even work?
We assign a timeout to null?

Reply:
> Does that even work?

Nope, it doesn't. It gives you a [CS0131](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0131)

Reply:
Right.

What's the original reason for adding the condition here, is it some sort of race condition with the connection shutting down (and setting WriteBuffer to null), and at the same time we set timeouts from replication?

If so, the code as is is still vulnerable to the race condition - `connector.WriteBuffer` may be set to null after the if checks for it, but before the timeout is set. The way to avoid this is to capture `connector.WriteBuffer` in a local variable, and then perform both the check and assignment on that variable.

Reply:
I've pushed this change so I can merge right away to continue work.

Reply:
> is it some sort of race condition with the connection shutting down (and setting WriteBuffer to null)

If I understand things correctly it's not a race condition but rather a case in normal linear code flow that we didn't exercise yet because we don't have tests for hard cancellation.

If we break the connector in the replication loop (`ReplicationConnection.StartReplicationInternal()`) because of a hard socket cancellation we immediately set the `WriteBuffer` property to `null` in `NpgsqlConnector.Cleanup()`. When we enter the finally block in `ReplicationConnection.StartReplicationInternal()` where we call `SetTimeouts()`, `NpgsqlConnector.WriteBuffer` is already set to `null`.

At some point we might want to add support for replication connections to your PostgreSQL mock implementation but for me today's not the day.